src/test/java/com/mojang/serialization/CodecTests.java

```diff
@@ -563,4 +563,41 @@ public void optionalField_lenientInvalidValues() {
             ))
         );
     }
+
+    private record Simple(String string, int integer) {
+        public static final Codec<Simple> CODEC = RecordCodecBuilder.create(i -> i.group(
+            Codec.STRING.fieldOf("string").forGetter(Simple::string),
+            Codec.INT.fieldOf("integer").forGetter(Simple::integer)
+        ).apply(i, Simple::new));
+    }
+
+    @Test
+    public void assumeMap_recordCodec() {
+        assertRoundTrips(
+            List.of(
+                Simple.CODEC,
+                // Wrapped directly
+                MapCodec.assumeMapUnsafe(Simple.CODEC).codec(),
+                // From a non-MapCodecCodec
+                MapCodec.assumeMapUnsafe(Codec.lazyInitialized(() -> Simple.CODEC)).codec()
```

Comment:
I'd prefer if it was extracted to method like "obfuscateType" to better describe what we are trying to achieve.
Also, `Codec.of(Simple.CODEC, Simple.CODEC)` would be probably simpler and less likely to accidentally do something unexpected